File: Java/135_Candy.java
public class Solution {
    public int candy(int[] ratings) {
        if(ratings==null||ratings.length==0)return 0;
        // 每个小孩至少有一个糖果，并且每个高优先级的小孩所拥有的糖果要多于他的低优先级的邻居
        // 要求总糖果数最少
        // 很自然地想到贪心法，怎么贪心？对于数列，想到要遍历，并且每次给予最少的糖果
        // 那么怎么满足题设条件？再次遍历，检查是否满足条件，如不满足，则给予最少糖果来满足条件
        int len=ratings.length;
        int[] candies = new int[len];
        candies[0]=1;
        // 第一遍从左往右，保证了所有左邻居满足题设条件
        for(int i=1;i<len;i++){
            if(ratings[i]>ratings[i-1]){
                candies[i]=candies[i-1]+1;
            }else{
                candies[i]=1;
            }
        }
        // 第二遍从右往左，保证所有右邻居也满足题设条件
        for(int i=len-2;i>=0;i--){
            if(ratings[i]>ratings[i+1]){
                candies[i]=Math.max(candies[i], candies[i+1]+1);
            }
        }
        int sum=0;
        for(int n:candies){
            sum+=n;
        }
        return sum;
    }
}

File: python/135_Candy.py
# coding=utf-8
import unittest

class Solution:
    def solve(self, nums):
        return self.candy(nums)

    # @param {integer[]} ratings
    # @return {integer}
    def candy(self, ratings):
        length=len(ratings)
        if length<2:
            return length
        candies=[1]*length
        for i in range(1,length):
            if ratings[i]>ratings[i-1]:
                candies[i] = candies[i-1]+1
        for i in range(length-2,-1,-1):
            candies[i]=max(candies[i+1]+1,candies[i]) if ratings[i]>ratings[i+1] else candies[i]
        return sum(candies)
        
class Test(unittest.TestCase):
    #override
    def setUp(self):
        self.s=Solution()
        
    def tearDown(self):
        print("TestComplete")
        
    def genTestCase(self):
        case=[
        [1,2,2],#4
        [1,2,1,3],#6
        [2,1],#3
        [3,2,1,2,2,1],#11
        [8,6,6,4,3,9,7,1],#15
        ]
        self.cases=case
        
    def test(self):
        self.genTestCase()
        for case in self.cases:
            print(("testcase: ", case))
            res=self.s.solve(case)
            print(("result: ",res))
            # if assert is needed
            
unittest.main()

